firebase-sessions/src/main/kotlin/com/google/firebase/sessions/SharedSessionRepository.kt

```diff
@@ -0,0 +1,237 @@
+        .catch {
+          val newSession =
+            SessionData(
+              sessionDetails = sessionGenerator.generateNewSession(null),
+              backgroundTime = null,
+            )
+          Log.d(
+            TAG,
+            "Init session datastore failed with exception message: ${it.message}. Emit fallback session ${newSession.sessionDetails.sessionId}",
+          )
+          emit(newSession)
+        }
+        .collect { sessionData ->
+          localSessionData = sessionData
+          val sessionId = sessionData.sessionDetails.sessionId
+          notifySubscribers(sessionId, NotificationType.GENERAL)
+        }
+    }
+  }
+
+  override fun appBackground() {
+    if (!::localSessionData.isInitialized) {
+      Log.d(TAG, "App backgrounded, but local SessionData not initialized")
+      return
+    }
+    val sessionData = localSessionData
+    Log.d(TAG, "App backgrounded on ${processDataManager.myProcessName} - $sessionData")
+
+    CoroutineScope(backgroundDispatcher).launch {
+      try {
+        sessionDataStore.updateData {
+          // TODO(mrober): Double check time makes sense?
+          sessionData.copy(backgroundTime = timeProvider.currentTime())
+        }
+      } catch (ex: Exception) {
+        Log.d(TAG, "App backgrounded, failed to update data. Message: ${ex.message}")
+        localSessionData = localSessionData.copy(backgroundTime = timeProvider.currentTime())
+      }
+    }
+  }
+
+  override fun appForeground() {
+    if (!::localSessionData.isInitialized) {
+      Log.d(TAG, "App foregrounded, but local SessionData not initialized")
+      return
+    }
+    val sessionData = localSessionData
+    Log.d(TAG, "App foregrounded on ${processDataManager.myProcessName} - $sessionData")
+
+    // Check if maybe the session data needs to be updated
+    if (isSessionExpired(sessionData) || isMyProcessStale(sessionData)) {
+      CoroutineScope(backgroundDispatcher).launch {
+        try {
+          sessionDataStore.updateData { currentSessionData ->
+            // Check again using the current session data on disk
+            val isSessionExpired = isSessionExpired(currentSessionData)
+            val isColdStart = isColdStart(currentSessionData)
+            val isMyProcessStale = isMyProcessStale(currentSessionData)
+
+            val newProcessDataMap =
+              if (isColdStart) {
+                // Generate a new process data map for cold app start
+                processDataManager.generateProcessDataMap()
+              } else if (isMyProcessStale) {
+                // Update the data map with this process if stale
+                processDataManager.updateProcessDataMap(currentSessionData.processDataMap)
+              } else {
+                // No change
+                currentSessionData.processDataMap
+              }
+
+            // This is an expression, and returns the updated session data
+            if (isSessionExpired || isColdStart) {
```

Comment:
From the dogfood today, when it is a cold start should we also update first session id in session data?